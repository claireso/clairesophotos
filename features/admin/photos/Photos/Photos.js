import { Fragment, useEffect, useCallback } from 'react'
import { useRouter } from 'next/router'

import PhotosReducer, { ACTION_TYPES } from '../reducer'

import Loader from '@components/Loader'
import { IconPlus } from '@components/Icons'
import { List, ListHeader } from '@components/List'
import { PrimaryButton } from '@components/Buttons'
import { Heading1 } from '@components/Headings'
import Pager from '@components/Pager'

import Photo from './components/Photo'

const Photos = () => {
  const [
    { items: photos, pager, ...state },
    { loadResources }
  ] = PhotosReducer.usePhotosReducer()
  const isLoading = ['idle', 'loading'].includes(state.status)

  const router = useRouter()
  const {
    query: { page },
    pathname
  } = router

  useEffect(() => {
    loadResources(page)
  }, [page, loadResources])

  const navigate = useCallback(
    (params) => {
      const query = {}
      if (page) query.page = page
      router.push({ pathname: pathname, query: { ...query, ...params } })
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [page]
  )

  const onPageChange = (page) => navigate({ page })

  const onCreate = useCallback(
    (ev) => {
      ev.preventDefault()
      navigate({ action: ACTION_TYPES.CREATE })
    },
    [navigate]
  )

  const onDelete = useCallback(
    (id, event) => {
      event.preventDefault()
      navigate({
        action: ACTION_TYPES.DELETE,
        id: id
      })
    },
    [navigate]
  )

  const onEdit = useCallback(
    (id, event) => {
      event.preventDefault()
      navigate({
        action: ACTION_TYPES.EDIT,
        id: id
      })
    },
    [navigate]
  )

  return (
    <Fragment>
      <ListHeader>
        <Heading1 data-testid="list-heading">
          Your photos {pager && <span>({pager.count})</span>}
        </Heading1>
        <PrimaryButton data-testid="button-create" onClick={onCreate}>
          Add a new photo
          <IconPlus />
        </PrimaryButton>
      </ListHeader>

      {isLoading ? (
        <Loader />
      ) : (
        <Fragment>
          <List>
            {photos.map((photo, index) => (
              <Photo
                key={index}
                {...photo}
                onDelete={onDelete}
                onEdit={onEdit}
              />
            ))}
          </List>
          <Pager {...pager} navigate={onPageChange} />
        </Fragment>
      )}
    </Fragment>
  )
}

export default Photos
