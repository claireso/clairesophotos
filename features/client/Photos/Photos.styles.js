import styled from 'styled-components'

export const PagerWrapper = styled.div`
  grid-column: 1 / -1;
  margin: 8.5rem 0;
`
