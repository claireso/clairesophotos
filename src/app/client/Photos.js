import React from 'react'
import PropTypes from 'prop-types'

import Photo from './components/Photo'
import Pager from './components/Pager'
import Button from './components/Button'

const Photos = (props = {}) => {
  return (
    <React.Fragment>
      {props.photos.map((photo, index) => <Photo key={index} {...photo} />)}
      <Pager
        {...props.pager}
        navigate={page => {
          const state = { page }
          window.history.pushState(state, '', `?page=${page}`)

          const popStateEvent = new PopStateEvent('popstate', { state })
          dispatchEvent(popStateEvent)
        }}
      >
        {({ items, getItemsProps }) => {
          return items.map((item, key) => (
            <li key={key}>
              <Button
                {...getItemsProps({
                  label: item.label,
                  title: item.title,
                  item: item
                })}
              />
            </li>
          ))
        }}
      </Pager>
    </React.Fragment>
  )
}

Photos.propTypes = {
  photos: PropTypes.array.isRequired,
  pager: PropTypes.object.isRequired
}

export default Photos
