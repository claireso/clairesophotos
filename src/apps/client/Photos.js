import React, { useCallback } from 'react'
import PropTypes from 'prop-types'

import Photo from './components/Photo'
import Pager from './components/Pager'

const Photos = (props = {}) => {
  const navigate = useCallback(page => {
    const state = { page }
    window.history.pushState(state, '', `?page=${page}`)

    const popStateEvent = new PopStateEvent('popstate', { state })
    dispatchEvent(popStateEvent)
  }, [])

  return (
    <React.Fragment>
      {props.photos.map((photo, index) => (
        <Photo key={index} {...photo} />
      ))}
      <Pager {...props.pager} navigate={navigate} />
    </React.Fragment>
  )
}

Photos.propTypes = {
  photos: PropTypes.array.isRequired,
  pager: PropTypes.object.isRequired
}

export default Photos
