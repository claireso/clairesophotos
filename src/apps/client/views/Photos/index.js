import React, { Fragment, useCallback } from 'react'
import PropTypes from 'prop-types'

import Photo from '../../components/Photo'
import Pager from '../../components/Pager'

const Photos = (props = {}) => {
  const navigate = useCallback(page => {
    const state = { page }
    window.history.pushState(state, '', `?page=${page}`)

    const popStateEvent = new PopStateEvent('popstate', { state })
    dispatchEvent(popStateEvent)
  }, [])

  return (
    <Fragment>
      {props.photos.map((photo, index) => (
        <Photo key={index} {...photo} row={index + 1} />
      ))}
      <Pager {...props.pager} navigate={navigate} />
    </Fragment>
  )
}

Photos.propTypes = {
  photos: PropTypes.array.isRequired,
  pager: PropTypes.object.isRequired
}

export default Photos
