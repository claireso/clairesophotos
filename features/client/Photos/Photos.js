import { Fragment, useCallback } from 'react'
import PropTypes from 'prop-types'
import Router from 'next/router'

import Pager from '@components/Pager'

import Photo from './components/Photo'

import * as S from './Photos.styles'

const Photos = (props = {}) => {
  const navigate = useCallback((page) => {
    Router.push({ pathname: '/', query: { page } })
    window.scrollTo(0, 0)
  }, [])

  return (
    <Fragment>
      {props.photos.map((photo, index) => (
        <Photo key={index} {...photo} row={index + 1} />
      ))}
      <S.PagerWrapper>
        <Pager {...props.pager} navigate={navigate} />
      </S.PagerWrapper>
    </Fragment>
  )
}

Photos.propTypes = {
  photos: PropTypes.array.isRequired,
  pager: PropTypes.object.isRequired
}

export default Photos
