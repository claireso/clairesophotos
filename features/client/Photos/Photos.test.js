import { render, fireEvent } from '@testing-library/react'
import Router from 'next/router'

import Photos from './index'

describe('<Photos />', () => {
  const renderComponent = () => {
    const { items: photos, pager } = global.__PHOTOS__

    return render(<Photos photos={photos} pager={pager} />)
  }

  test('should render component', () => {
    const { container } = renderComponent()

    expect(container).toMatchSnapshot()
  })

  test('should navigate', () => {
    Router.push = jest.fn()

    const { getByTitle } = renderComponent()

    fireEvent.click(getByTitle('Next page'))

    expect(Router.push).toHaveBeenCalledWith({
      pathname: '/',
      query: { page: 2 }
    })

    fireEvent.click(getByTitle('Last page'))

    expect(Router.push).toHaveBeenCalledWith({
      pathname: '/',
      query: { page: 19 }
    })
  })
})
