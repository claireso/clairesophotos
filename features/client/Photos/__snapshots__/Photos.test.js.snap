// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Photos /> should render component 1`] = `
.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  list-style-type: none;
  width: 100%;
  padding: 0;
}

.c6 > li {
  margin: 0 0.4rem;
}

.c8 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: none;
  border: none;
  color: currentColor;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji;
  outline: none;
  padding: 0;
  margin: 0;
  border-radius: 0.2rem;
  font-size: 1.2rem;
  line-height: 1.3333333;
  padding: 1.6rem;
  -webkit-transition: background 150ms ease-out;
  transition: background 150ms ease-out;
  background: var(--secondary-normal);
  color: var(--text-normal);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 1.4rem;
  height: 4.4rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0;
  width: 4.4rem;
}

.c8 > svg {
  margin: 0 -0.4rem 0 0.4rem;
}

.c8 + .c7 {
  margin: 0 0 0 0.8rem;
}

.c8:hover {
  background: var(--secondary-darker);
}

.c8 + .Buttonsstyles__PrimaryButton-jpizpn-2 {
  margin-left: 0.8rem;
}

.c0 {
  grid-row-start: 1;
  grid-column: 1 / span 6;
  margin: 0;
}

.c4 {
  grid-row-start: 2;
  grid-column: 1 / span 6;
  margin: 0;
}

.c2 {
  font-size: 1.3rem;
}

.c3 {
  color: var(--gray-normal);
  display: block;
  font-size: 1.1rem;
  font-style: italic;
  margin: 0.4rem 0 0;
}

.c1 {
  background: var(--gray-lighter);
  position: relative;
  padding-top: calc(0.666667 * 100%);
  width: 100%;
  margin: 0 0 1rem;
}

.c5 {
  grid-column: 1 / -1;
  margin: 8.5rem 0;
}

@media (min-width:800px) {
  .c0 {
    grid-column: calc(var(--grid-number-column-large) - 8) / span 8;
  }

  .c0 + figure {
    margin-top: 20rem;
  }
}

@media (min-width:800px) {
  .c4 {
    grid-column: 3 / span 8;
  }

  .c4 + figure {
    margin-top: 20rem;
  }
}

<div>
  <figure
    class="c0"
  >
    <div
      class="c1"
    />
    <figcaption
      class="c2"
    >
      <span />
      <span
        class="c3"
      >
        Février 2019
      </span>
    </figcaption>
  </figure>
  <figure
    class="c4"
  >
    <div
      class="c1"
    />
    <figcaption
      class="c2"
    >
      <span />
      <span
        class="c3"
      >
        Janvier 2019
      </span>
    </figcaption>
  </figure>
  <div
    class="c5"
  >
    <ul
      class="c6"
    >
      <li>
        <button
          class="Buttonsstyles__DefaultButton-jpizpn-0 Buttonsstyles__StyledButton-jpizpn-1 Buttonsstyles__SecondaryButton-jpizpn-4 c7 c8"
          title="Next page"
        >
          »
        </button>
      </li>
      <li>
        <button
          class="Buttonsstyles__DefaultButton-jpizpn-0 Buttonsstyles__StyledButton-jpizpn-1 Buttonsstyles__SecondaryButton-jpizpn-4 c7 c8"
          title="Last page"
        >
          »»
        </button>
      </li>
    </ul>
  </div>
</div>
`;
